### PR TITLE
[Feature] Deep-linking changes to view `Params` and `primitive.cast` 

### PR DESCRIPTION
**Pull request by Manrich121**
[ch12166](https://app.shortcut.com/journeyapps/story/12166)

Features and changes needed for deep-linking work:
- `required` and `provide-value` to view `params` 
- Update `cast` of primitives to support `string` -> `type`

### Note
This branched off from https://github.com/journeyapps-platform/journey-lib/pull/10 and will only introduce the changes needed for deep-linking.

### CODE REVIEW

File: journeyapps-db/src/types/QueryType.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    return query._clone();
+  };
+  cast = function (this: QueryType, value: any) {
+    if (typeof value != 'object') {
+      throw new Error(value + ' is not an object');
+    }
+    const thisTypeName = this.objectType.name;
+    if (
+      value.type != null &&
+      value.type instanceof ObjectType &&
```
Comment (dylanvorster) [inline journeyapps-db/src/types/QueryType.ts:17]: possible copy paste of this entire function? should probably be `Query` and  as well as above on like 7

File: journeyapps-db/src/types/primitives/Datetime.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+        return null;
+      } else {
+        return new Date(parsed);
+      }
+    } else {
+      return null;
+    }
+  }
+
+  format(value: any, format: string = 'MMMM D YYYY h:mm A'): string {
```
Comment (dylanvorster) [inline journeyapps-db/src/types/primitives/Datetime.ts:27]: would be better to put this in a static variable in this class called FORMAT_DEFAULT

File: journeyapps-db/src/types/primitives/Location.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
+import { LocationType as SchemaLocationType } from '@journeyapps/parser-schema';
+import { Location } from '../Location';
+import { DBTypeMixin } from '../Type';
+
+export class LocationType extends DBTypeMixin(SchemaLocationType) {
+  valueFromJSON(value: any) {
+    if (Array.isArray(value)) {
```
Comment (dylanvorster) [inline journeyapps-db/src/types/primitives/Location.ts:7]: probably a good idea to check the length of this array else, throw an error

**Review by dylanvorster — DISMISSED** [review]
looks good, just worried about a possible copy paste

**Review by wessels2 — APPROVED** [review]

**Pull request merged**